File: Java/Dialogue.java
import java.util.List;

public final class Dialogue {
    private final List<DialogueSteps.IStep> steps;
    private final Runnable onEnd;

    Dialogue(List<DialogueSteps.IStep> steps, Runnable onEnd) {
        this.steps = List.copyOf(steps);
        this.onEnd = onEnd;
    }

    public static DialogueBuilder create() {
        return new DialogueBuilder();
    }

    public List<DialogueSteps.IStep> steps() {
        return steps;
    }

    public DialogueResult play(DialogueRenderer renderer) {
        return new DialogueRuntime(renderer).execute(this);
    }

    void end() {
        if (onEnd != null) onEnd.run();
    }
}

File: C#/Dialogue.cs
namespace DialogueSystem;

public sealed class Dialogue
{
    public IReadOnlyList<IDialogueStep> Steps { get; }
    private readonly Action? onEnd;

    internal Dialogue(IReadOnlyList<IDialogueStep> steps, Action? onEnd)
    {
        this.Steps = steps;
        this.onEnd = onEnd;
    }

    public static DialogueBuilder Create() => new DialogueBuilder();
    public DialogueResult Play(IDialogueRenderer renderer) => new DialogueRuntime(renderer).Execute(this);

    internal void TriggerEnd() => onEnd?.Invoke();
}
